src/main/java/org/roaringbitmap/longlong/Roaring64NavigableMap.java

```diff
@@ -0,0 +1,1351 @@
+      } else {
+        previousBucketCardinality = sortedCumulatedCardinality[highPosition - 1];
+      }
+
+      BitmapDataProvider lowBitmap = highToBitmap.get(sortedHighs[highPosition]);
+
+      // Rank is previous cardinality plus rank in current bitmap
+      return previousBucketCardinality + lowBitmap.rankLong(low);
+    } else {
+      // There is no bucket holding this item: insertionPoint is previous bitmap
+      int insertionPoint = -highPosition - 1;
+
+      if (insertionPoint == 0) {
+        // this key is before all inserted keys
+        return 0;
+      } else {
+        // The rank is the cardinality of this previous bitmap
+        return sortedCumulatedCardinality[insertionPoint - 1];
+      }
+    }
+  }
+
+  // https://github.com/RoaringBitmap/CRoaring/blob/master/cpp/roaring64map.hh
+  private long rankLongNoCache(int high, int low) {
+    long result = 0L;
+
+    BitmapDataProvider lastBitmap = highToBitmap.get(high);
+    if (lastBitmap == null) {
+      // There is no value with same high: the rank is a sum of cardinalities
+      for (Entry<Integer, BitmapDataProvider> bitmap : highToBitmap.entrySet()) {
+        if (bitmap.getKey().intValue() > high) {
+          break;
+        } else {
+          result += bitmap.getValue().getLongCardinality();
+        }
+      }
+    } else {
+      for (BitmapDataProvider bitmap : highToBitmap.values()) {
+        if (bitmap == lastBitmap) {
+          result += bitmap.rankLong(low);
+          break;
+        } else {
+          result += bitmap.getLongCardinality();
+        }
+      }
+    }
+
+    return result;
+  }
+
+  /**
+   * 
+   * @param high for which high bucket should we compute the cardinality
+   * @return the highest validatedIndex
+   */
+  protected int ensureCumulatives(int high) {
+    if (allValid) {
+      // the whole array is valid (up-to its actual length, not its capacity)
+      return highToBitmap.size();
+    } else if (compare(high, firstHighNotValid) < 0) {
+      // The high is strictly below the first not valid: it is valid
+
+      // sortedHighs may have only a subset of valid values on the right. However, these invalid
+      // values have been set to maxValue, and we are here as high < firstHighNotValid ==> high <
+      // maxHigh()
+      int position = binarySearch(sortedHighs, high);
+
+      if (position >= 0) {
+        // This high has a bitmap: +1 as this index will be used as right (excluded) bound in a
+        // binary-search
+        return position + 1;
+      } else {
+        // This high has no bitmap: it could be between 2 highs with bitmaps
+        int insertionPosition = -position - 1;
+        return insertionPosition;
+      }
+    } else {
+      // For each deprecated buckets
+      SortedMap<Integer, BitmapDataProvider> tailMap =
+          highToBitmap.tailMap(firstHighNotValid, true);
+
+      // TODO .size on tailMap make an iterator: arg
+      int indexOk = highToBitmap.size() - tailMap.size();
```

Comment:
Cannot you look for the index of firstHighNotValid instead, if you want to avoid the call the other iterator.